### Cramér's V

[link to video](https://youtube.com/shorts/F-071xdEwQg?si=mrnbmtqNFqKV2v_u)

In [28]:
import pandas as pd
df = pd.read_csv("/home/omkar/Omkar Pawar/Data Science/Data Decides Code for YouTube Shorts/Datasets/CustomerPreferencesDataset.csv")
df

CustomerID SubscriptionType PreferredGenre PaymentMethod      Device
0           1            Basic          Drama   Credit Card    Smart TV
1           2          Premium         Comedy        PayPal      Tablet
2           3         Standard         Action    Debit Card  Smartphone
3           4          Premium          Drama        PayPal      Laptop
4           5            Basic         Comedy    Debit Card  Smartphone
5           6          Premium         Action   Credit Card    Smart TV
6           7         Standard         Comedy    Debit Card      Laptop
7           8            Basic          Drama   Credit Card  Smartphone
8           9          Premium         Action        PayPal    Smart TV
9          10            Basic         Comedy   Credit Card      Tablet

In [33]:
from scipy.stats import chi2_contingency

def cramers_v_test(data, feature1, feature2):
    conti ngency_table = pd.crosstab(data[feature1], data[feature2])
    chi2, _, _, _ = chi2_contingency(contingency_table)
    n = min(contingency_table.shape)
    cramers_v = (chi2 / (len(data) * (n - 1))) ** 0.5
    return cramers_v

#------------------------------------------------------------ 
# separating categorical columns from numerical columns: 
categorical_columns = df.select_dtypes(include=['object']).columns

for i in range(len(categorical_columns)):
    for j in range(i + 1, len(categorical_columns)):
        feature1 = categorical_columns[i]
        feature2 = categorical_columns[j]
        result = cramers_v_test(df, feature1, feature2)
        print(f"Cramér's V for {feature1} and {feature2}: {result}")

Cramér's V for SubscriptionType and PreferredGenre: 0.4208127057650866
Cramér's V for SubscriptionType and PaymentMethod: 0.75
Cramér's V for SubscriptionType and Device: 0.5204164998665333
Cramér's V for PreferredGenre and PaymentMethod: 0.33850160019316505
Cramér's V for PreferredGenre and Device: 0.5713045500334204
Cramér's V for PaymentMethod and Device: 0.5464532103585001


- 0.0 to 0.1: Very weak association
- 0.1 to 0.2: Weak association
- 0.2 to 0.3: Moderate association
- 0.3 to 0.4: Moderately strong association
- 0.4 to 0.5: Strong association
- 0.5 and above: Very strong association

Here's an interpretation for each result:

1. **SubscriptionType and PreferredGenre (0.4208):**
   - Moderate association: There is a moderate relationship between a customer's subscription type and their preferred genre. The Cramér's V value of 0.4208 indicates a moderate strength of association.

2. **SubscriptionType and PaymentMethod (0.75):**
   - Strong association: There is a strong relationship between a customer's subscription type and their payment method. The high Cramér's V value of 0.75 suggests a robust association.

3. **SubscriptionType and Device (0.5204):**
   - Moderate association: There is a moderate relationship between a customer's subscription type and the device they use for streaming. The Cramér's V value of 0.5204 indicates a moderate strength of association.

4. **PreferredGenre and PaymentMethod (0.3385):**
   - Weak to Moderate association: There is a weak to moderate relationship between a customer's preferred genre and their payment method. The Cramér's V value of 0.3385 suggests a modest strength of association.

5. **PreferredGenre and Device (0.5713):**
   - Moderate to Strong association: There is a moderate to strong relationship between a customer's preferred genre and the device they use for streaming. The Cramér's V value of 0.5713 indicates a notable strength of association.

6. **PaymentMethod and Device (0.5465):**
   - Moderate to Strong association: There is a moderate to strong relationship between a customer's payment method and the device they use for streaming. The Cramér's V value of 0.5465 suggests a significant strength of association.

These interpretations are based on the magnitude of the Cramér's V values, where **higher values indicate stronger associations**. It's essential to consider these results in the context of your specific dataset and the variables being analyzed.

### Additional Resources:
- https://www.statology.org/interpret-cramers-v/
- https://www.statology.org/cramers-v-in-python/